### ISSUE TITLE
Training Kraken, am I doing something wrong?

### ISSUE THREAD

**Issue by christophered** [mittagessen/kraken#44]
[A video demonstration of the problem I am facing while training a new CLSTM model.](https://youtu.be/HF0BNSjgZhA)

I have transcribed and extracted a sample Arabic image of 7 lines, just for testing, [see attachment.](https://github.com/mittagessen/kraken/files/1049616/sample.zip)

I have used [train.sh from mittagessen/kraken-vagrant](https://github.com/mittagessen/kraken-vagrant/blob/master/training/train.sh).

I have tested training using train.sh setting values to 1000, and second test set to 10.
![untitled4](https://cloud.githubusercontent.com/assets/16248376/26754024/83bbcc42-487b-11e7-8df0-27048e0d4abc.png)

 

- In both occasions, train.sh started creating multiple arabic-*.clstm which all had the same file size.

- I have used one of the .clstm files in kraken recognition, the result was an empty .txt file with only 6 lines of empty space.

- I have downloaded a kraken model just to test if the recognition function is working, and yes its working.

Am I doing something wrong here?
Waiting for your reply

**Comment by mittagessen** [comment]
Models are always the same size. Memory-wise they are just fixes-size matrices and training is adjusting the values until they start recognizing stuff. As training takes quite a while (~8-12h or ~30000 iterations) it is entirely possible that you just haven't trained long enough. The output of an untrained network will be mostly whitespace, over time it learns characters. You can get a rough estimation from the output of the train.sh script although the order of characters will be reversed.

**Comment by christophered** [comment]
If I generate an arabic.clstm using clstmocrtrain, can I use kraken later to continue training?
ketos train -l?
what command should i write? how?

**Comment by mittagessen** [comment]
yes. ```ketos train --no-reorder -i ~/model.clstm data/*.png``` should do the trick (as you already reordered the code points to train with clstmocrtrain.

**Comment by christophered** [comment]
This error appears:

vagrant@kraken:~/Desktop/sample/output$ ketos train --no-reorder -i arabic-8000.clstm *.png
Building ground truth set	⣷
✓
Loading model	⣾Traceback (most recent call last):
  File "/usr/local/bin/ketos", line 11, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kraken/ketos.py", line 113, in train
    rnn = models.ClstmSeqRecognizer(load)
  File "/usr/local/lib/python2.7/dist-packages/kraken/lib/models.py", line 43, in __init__
    import clstm
ImportError: No module named clstm

**Comment by christophered** [comment]
I also tried adding this code to train.sh and executing it, but no use, it starts from beginning 0
load=arabic-8000.clstm
start=8000

**Comment by mittagessen** [comment]
The first error (ImportError) is because the python clstm bindings aren't installed. For the second one I believe it's just the counter in the output starting from zero.

**Comment by andbue** [comment]
Hi there,
I made it work yesterday with https://github.com/tmbdev/clstm, but I had to change some lines in model.py (https://github.com/mittagessen/kraken/compare/master...andbue:master).
Cheers
Andreas

**Comment by christophered** [comment]
@mittagessen 
> the python clstm bindings aren't installed

Do you mean that it is not currently possible to continue training using kraken?

**Comment by christophered** [comment]
@mittagessen what about @andbue  other commit
[Adapted to current clstm code](https://github.com/mittagessen/kraken/commit/dd7eb3e4978409c2714bebb1875d678252f61879)

**Comment by mittagessen** [comment]
Uhhm are you sure you installed the ``separate_derivs`` branch of CLSTM? The python bindings don't work for master and the ``separate_derivs`` python bindings do not have for example the ``sgd_update`` method called by @andbue's changes.

**Comment by christophered** [comment]
@mittagessen i am currently testing training [the Arabic golden slandered data,](https://github.com/OpenArabic/OCR_GS_Data/tree/master/ara/book_Yacqubi.Tarikh/7_final) but the text in their .txt is written in  RTL, how can I flip it to be in LTR reorder and normalize as similar to what is Kraken doing in the extract?
Because I was training in clstmocrtrain and reached 93,000 epochs with 100% error rate and 0 OUT, then I noticed the problem that the .txt needs to be Reordered and normalized.
How can I do that on an already available .txt
Waiting for your reply eagerly

**Comment by andbue** [comment]
@mittagessen I didn't know about the separate_derivs branch, so I built and installed master. The python bindings work sufficiently well for kraken. The only thing that didn't work for me was creating new networks from python. As clstm seems to be no longer actively developed, maybe it could be an option to use pytorch or something similar in the future?

**Comment by mittagessen** [comment]
Interesting, last time I tried they didn't even compile. I had a look at most machine learning frameworks with python interfaces (caffe, tensorflow, theano). They all suffer from 2 problems to some extent: network serialization is unclean (pickle ffs) and model instantiation is extremely slow because of nontrivial load-time optimization. Even when caching compilation artifacts instantiating a model will always take upwards of a minute which is fine if you're doing research or keep them all in memory but it is somewhat annoying for doing real OCR.

I've got some plug-in keras code that could form a new recognition backend that somewhat solves the first issue but compile-time still feels like a step back, especially as it doesn't perform significantly better on CPU than CLSTM (and uses a lot more memory).
